File: 2025/Day2/D2PartTwo.py

def get_file_data(filename:str = "sample.txt") -> list[list[int]]:
    with open(filename, 'r') as f:
        d = f.read().split("\n")[0].split(",")
    for i in range(len(d)):
        d[i] = d[i].split("-")
        d[i] = [int(d[i][0]), int(d[i][1])]
    return d

def has_repeating_digits(numb:int) -> bool:
    def convert_to_int(st:str) -> int:
        res = []
        for ch in st:
            res.append(int(ch))
        return res
    
    numb_str = str(numb)
    numb_list = convert_to_int(numb_str)
    
    cycle = []
    for i in range(1, len(numb_list)):
        if len(numb_list) % i == 0:
            cycle = numb_list[:i]
            if len(cycle) < 1:
                continue
            found_repeating = True
            for i in range(len(numb_list)):
                if cycle[i % len(cycle)] != numb_list[i]:
                    found_repeating = False
                    break
            if found_repeating:
                return True
    return False

def calculate_invalid_ids(start:int, end:int) -> list[int]:
    total = []
    for i in range(start, end+1):
        if has_repeating_digits(i):
            total.append(i)
    return total

def iterate_ranges(ranges:list[list[int]]) -> int:
    total = 0
    for r in ranges:
        resp = calculate_invalid_ids(r[0], r[1])
        print(f"For range {r}, received responses {resp}")
        total += sum(resp)
    return total

def test_system_test_the_thing():
    file_data = get_file_data()
    resp = iterate_ranges(file_data)
    print("Test RESPONSE:", resp)
    assert resp == 4174379265

def test_121121():
    resp = iterate_ranges([[121120, 121122]])
    print("Test 121121 response:", resp)
    assert resp == 121121

if __name__ == "__main__":
    test_121121()
    print("Test 121121 passed.")

    test_system_test_the_thing()
    print("System/Function Test Passed.")

    file_data = get_file_data(filename="input.txt")
    file_data.sort()
    print("File_data:", file_data)
    resp = iterate_ranges(file_data)
    print("Test:", resp)
    if resp != 25663320831:
        print("\nDIFF:", abs(resp - 25663320831))
    assert resp == 25663320831

File: 2025/Day2/D2PartTwo.swift
import Foundation

func getFileData(filename:String = "sample.txt") throws -> [[Int]] {
    let fileRead = try String(contentsOfFile: filename, encoding: String.Encoding.utf8)
                    .split(separator: ",")
                    .map {
                        String($0).last != "\n" ? String($0) : String(String($0).dropLast(1))
                    }
    let res:[[Int]] = fileRead.map {
                    let temp = $0.split(separator: "-")
                    return [Int(temp[0])!, Int(temp[1])!]
                }
    return res
}

func hasRepeatingDigits(_ numb:Int) -> Bool {
    let numbStr = String(describing: numb) //This could also be "\(numb)"
    if numbStr.count % 2 == 1 { //Why bother if you can't repeat numbers anyways?
        return false
    }
    var cycle:[Int] = []
    var current:Int = 0
    var position:String.Index = numbStr.startIndex
    for i in 0..<numbStr.count {
        position = numbStr.index(numbStr.startIndex, offsetBy: i)
        let cyclePosition = cycle.index(cycle.startIndex, offsetBy: i % cycle.count)
        print(".> \(cycle[cyclePosition]) \(numbStr[position])")
        if cycle.count == 0 || (numbStr.count / 2 > cycle.count && cycle[cyclePosition] == Int(numbStr[position])) {
            cycle = numbStr[numbStr.startIndex...position].map { Int(String($0))! }
        } else {
            return false
        }
    }
    return true
}

func calculateInvalidIDs(_ start:Int, _  end:Int) -> Int {
    var total:Int = 0
    for i in start...end {
        if hasRepeatingDigits(i) {
            print("> \(i) has repeating numbers")
            total += i
        }
    }
    return total
}

func iterRanges(_ ranges:[[Int]]) -> Int {
    var total:Int = 0
    for range in ranges {
        total += calculateInvalidIDs(range[0], range[1])
    }
    return total
}

let fileData:[[Int]] = try getFileData()
print("Test getFileData(): \(fileData)")
print()
print("Test 1188511885: \(hasRepeatingDigits(1188511885))")
print()
// print("Invalid ID Sum: \(iterRanges(fileData))")
// print("Test hasRepeatingDigits: \(hasRepeatingDigits(124232124232))")